### PR TITLE
Send pending submission to RWS

### PR DESCRIPTION
**Pull request by myungwoo**
I'm considering the CMS for ICPC.

It is the first step for it.
Almost all ICPC scoreboard needs to display pending(not evaluated yet) submission also.
Submission is considered as pending in RWS when it has no subchange(score data) but submission.

+)
My idea for CMS - ICPC is, the submission is accepted when it got full score,
no special setting in CMS (new scoring method, penalty etc...),
just send RWS data to already made ICPC ScoreBoard([spotboard](https://github.com/wookayin/spotboard), [spotboard feedserver for cms](https://github.com/myungwoo/spotboard-cms)).

---

This change is [Reviewable](https://reviewable.io/reviews/cms-dev/cms/610)

### CODE REVIEW

File: cms/service/ProxyService.py [hunk 1]
```diff
@@ -448,6 +476,34 @@ def reinitialize(self):
...
+        Usually called by EvaluationService when it's done with submitting,
+        and then send data about the submission to the rankings.
+
+        submission_id (int): the id of the submission that changed.
+
+        """
+        with SessionGen() as session:
+            submission = Submission.get_from_id(submission_id, session)
+
+            if submission is None:
```
Comment (stefano-maggiolo) [inline cms/service/ProxyService.py]: Codacy Issue found: [Line too long (81/80)](https://www.codacy.com/app/s-maggiolo/cms/file/2605456822/issues/source?bid=3167167&fileBranchId=3309490#l492)

Comment (myungwoo) [inline cms/service/ProxyService.py]: Done

File: cms/service/ProxyService.py [hunk 2]
```diff
@@ -369,6 +377,26 @@ def initialize(self):
         self.enqueue(ProxyOperation(ProxyExecutor.USER_TYPE, users))
         self.enqueue(ProxyOperation(ProxyExecutor.TASK_TYPE, tasks))
```
Comment (stefano-maggiolo) [inline cms/service/ProxyService.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/s-maggiolo/cms/file/2607019884/issues/source?bid=3167167&fileBranchId=3309490#l380)

Comment (myungwoo) [inline cms/service/ProxyService.py]: Done.

**Comment by myungwoo** [comment]
May I know the opinions why this PR is not accepted yet?

**Comment by lerks** [comment]
I want to take a look at it to make sure that all cases are correctly handled (RWS is a pretty messy tangle of code), and to maybe take this chance to clean up a little. And for now I haven't had the time to do that, sorry. I'm still finishing the fix for #612.

BTW, I didn't really understand why a scoreboard for the ICPC has to know about submissions before they're graded. (I've taken part in the ICPC but I've never looked at a live scoreboard, so I don't know what is shown there). What's wrong with sending the information about the submission once it scored?

**Comment by myungwoo** [comment]
There's nothing wrong without sending submission before it scored.

However, there's a scoreboard which supports displaying pending submissions. And I think it is pretty good because it contains more information of contest and sometimes evaluation takes too much time, so it is good to let audience know that submissions are exist even if it's not evaluated yet.

**Comment by lerks** [comment]
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[cms/service/ProxyService.py, line 291 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/610#-KKi0wrqbsnoe7hRP27-:-KKi0wrqbsnoe7hRP270:-560717049) ([raw file](https://github.com/cms-dev/cms/blob/2301fde751a99009e660f0143a68e857ef88a397/cms/service/ProxyService.py#L291)):*
> ```Python
>                 # We need to send data of submission which is
>                 # not evaluated yet to RWS.
>                 for operation in self.operations_for_submission(submission):
> ```

This makes all submissions show up as missing operations at every sweep, causing them to be sent again. It's undesirable. Please put in place a system like `scores_sent_to_ranking` and `tokens_sent_to_ranking`.

---

*[cms/service/ProxyService.py, line 380 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/610#-KKi1jM4QMgrtU8yW6Y-:-KKi1jM5OpcYixBazzZr:-355563727) ([raw file](https://github.com/cms-dev/cms/blob/2301fde751a99009e660f0143a68e857ef88a397/cms/service/ProxyService.py#L380)):*
> ```Python
>         self.enqueue(ProxyOperation(ProxyExecutor.TASK_TYPE, tasks))
> 
>     def operations_for_submission(self, submission):
> ```

Could you have `operations_for_score` and `operations_for_token` call this function, to avoid code duplication?

---

*[cms/service/ProxyService.py, line 492 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/610#-KKi2CgDmVJd8kSw3zbu:-KKi2CgDmVJd8kSw3zbv:-1290967817) ([raw file](https://github.com/cms-dev/cms/blob/2301fde751a99009e660f0143a68e857ef88a397/cms/service/ProxyService.py#L492)):*
> ```Python
> 
>             if submission is None:
>                 logger.error("[submission_submitted] Received score request for "
> ```

You copied and pasted these messages. Please modify them.

---

*[cms/service/ProxyService.py, line 497 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/610#-KKi2JNNXqidFoC_SHqc:-KKi2JNNXqidFoC_SHqd:-918288111) ([raw file](https://github.com/cms-dev/cms/blob/2301fde751a99009e660f0143a68e857ef88a397/cms/service/ProxyService.py#L497)):*
> ```Python
> 
>             if submission.participation.hidden:
>                 logger.info("[submission_submitted] Score for submission %d "
> ```

This one too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/610)*

**Comment by lerks** [comment]
I'm not done with the review yet.

**Comment by myungwoo** [comment]
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[cms/service/ProxyService.py, line 291 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/610#-KKi0wrqbsnoe7hRP27-:-KKpwROVhqCNvcskLzRY:-1347246880) ([raw file](https://github.com/cms-dev/cms/blob/2301fde751a99009e660f0143a68e857ef88a397/cms/service/ProxyService.py#L291)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary>
> This makes all submissions show up as missing operations at every sweep, causing them to be sent again. It's undesirable. Please put in place a system like `scores_sent_to_ranking` and `tokens_sent_to_ranking`.
</details>
I already use `submissions_sent_to_ranking`, could you please check this part again?

line 291

---

*[cms/service/ProxyService.py, line 380 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/610#-KKi1jM4QMgrtU8yW6Y-:-KKsRblx7wJL-wy-UrGa:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/2301fde751a99009e660f0143a68e857ef88a397/cms/service/ProxyService.py#L380)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary>
> Could you have `operations_for_score` and `operations_for_token` call this function, to avoid code duplication?
</details>
Done.

---

*[cms/service/ProxyService.py, line 492 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/610#-KKi2CgDmVJd8kSw3zbu:-KKsRnkRu7d2X__Vf9Ou:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/2301fde751a99009e660f0143a68e857ef88a397/cms/service/ProxyService.py#L492)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary>
> You copied and pasted these messages. Please modify them.
</details>
Done.

---

*[cms/service/ProxyService.py, line 497 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/610#-KKi2JNNXqidFoC_SHqc:-KKsRoDPIcIO_g5CspIC:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/2301fde751a99009e660f0143a68e857ef88a397/cms/service/ProxyService.py#L497)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary>
> This one too.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/610)*

**Comment by wil93** [comment]
@myungwoo why was this closed? It was almost merged :cry:

**Comment by myungwoo** [comment]
@wil93 I closed it because I didn't have enough time at that moment, and that's still true now. 😢

**Pull request closed without merging**